docs/architecture/dapr-for-net-developers/publish-subscribe-buildingblock.md

```diff
@@ -0,0 +1,415 @@
+
+- When a user checks-out a shopping basket.
+- When a payment for an order has succeeded.
+- When the grace-period of a purchase has expired.
+
+In the earlier eShopOnContainers, the Publish/Subscribe pattern was implemented using an *IEventBus* abstraction:
+
+```csharp
+public interface IEventBus
+{
+    void Publish(IntegrationEvent @event);
+
+    void Subscribe<T, TH>()
+        where T : IntegrationEvent
+        where TH : IIntegrationEventHandler<T>;
+}
+```
+
+Multiple concrete implementations were built for this abstraction,  including RabbitMQ and Azure Service Bus. Each implementation includes a great deal of custom plumbing code that was complex and difficult to maintain.
+
+### Publishing events
+
+In the updated eShopOnDapr, we replaced the entire Publish/Subscribe implementation with Dapr. This refactoring was straightforward. We minimized the existing `IEventBus` abstraction, so it only contains a method for publishing events:
+
+```csharp
+public interface IEventBus
+{
+    Task PublishAsync(IntegrationEvent @event);
+}
+```
+
+To implement the `IEventBus` abstraction, a single `DaprEventBus` implementation was created. It uses the Dapr client to publish an event:
+
+```csharp
+public class DaprEventBus : IEventBus
+{
+    private const string DAPR_PUBSUB_NAME = "pubsub";
+
+    private readonly DaprClient _dapr;
+    private readonly ILogger<DaprEventBus> _logger;
+
+    public DaprEventBus(DaprClient dapr, ILogger<DaprEventBus> logger)
+    {
+        _dapr = dapr;
+        _logger = logger;
+    }
+
+    public async Task PublishAsync<TIntegrationEvent>(TIntegrationEvent @event)
+        where TIntegrationEvent : IntegrationEvent
+    {
+        var topicName = @event.GetType().Name;
+
+        _logger.LogInformation("Publishing event {@Event} to {PubsubName}.{TopicName}", @event, DAPR_PUBSUB_NAME, topicName);
+
+        // We need to make sure that we pass the concrete type to PublishEventAsync,
+        // which can be accomplished by casting the event to dynamic. This ensures
+        // that all event fields are properly serialized.
+        await _dapr.PublishEventAsync(DAPR_PUBSUB_NAME, topicName, (dynamic)@event);
+    }
+}
+```
+
+As you can see in the code snippet, the topic name is derived from event type's name. Because all eShop services use the `IEventBus` abstraction, retrofitting Dapr required *absolutely no change* to the mainline application code.
+
+> Note how we cast the event parameter to a C# `dynamic` type. This workaround addresses an issue found in the `System.Text.Json` serializer, the Dapr SDK uses to serialize/deserialize messages. In the eShop code, an event is sometimes explicitly declared as the base-class for events `IntegrationEvent`. This is done because the specific event-type is determined dynamically at runtime based on business-logic, as seen in the following example:
```

Comment:
We shouldn't call this an issue with S.T.Json. In .NET Core 3.1, S.T.Json has no real support for polymorphism. Let's say we've got a base class called `Shape` and a concrete class called `Circle` derived from `Shape`.
Calling `Serialize<Shape>(aCircleInstance)` will only serialize the members of `Shape` and ignore any members defined on `Circle`. This is by design. Using "issue" makes it look like a bug.

Reply:
I've updated the text accordingly.